core/src/main/java/org/pentaho/di/core/database/DatabaseMetaInformation.java

```diff
@@ -0,0 +1,453 @@
+            while ( catalogTablesResultSet.next() ) {
+              String tableName = catalogTablesResultSet.getString( 3 );
+
+              if ( !db.isSystemTable( tableName ) ) {
+                catalogTables.add( tableName );
+              }
+            }
+            // Immediately close the catalog tables ResultSet
+            //
+            catalogTablesResultSet.close();
+
+            // Sort the tables by names
+            Collections.sort( catalogTables );
+          } catch ( Exception e ) {
+            // Obviously, we're not allowed to snoop around in this catalog.
+            // Just ignore it!
+            // LogWriter.getInstance().logError(getClass().getName(),BaseMessages.getString(PKG,
+            // "DatabaseMeta.Error.UnexpectedCatalogError"), e);
+          }
+
+          // Save the list of tables in the catalog (can be empty)
+          //
+          catalog.setItems( catalogTables.toArray( new String[catalogTables.size()] ) );
+        }
+
+        // Save for later...
+        setCatalogs( catalogList.toArray( new Catalog[catalogList.size()] ) );
+      }
+      if ( monitor != null ) {
+        monitor.worked( 1 );
+      }
+
+      if ( monitor != null && monitor.isCanceled() ) {
+        return;
+      }
+      if ( monitor != null ) {
+        monitor.subTask( BaseMessages.getString( PKG, "DatabaseMeta.Info.GettingSchemaInfo" ) );
+      }
+      if ( databaseMeta.supportsSchemas() && dbmd.supportsSchemasInTableDefinitions() ) {
+        ArrayList<Schema> schemaList = new ArrayList<Schema>();
+        try {
+          String schemaFilterKey = databaseMeta.getPluginId() + "." + FILTER_SCHEMA_LIST;
+          if ( ( connectionExtraOptions != null ) && connectionExtraOptions.containsKey( schemaFilterKey ) ) {
+            String schemasFilterCommaList = connectionExtraOptions.get( schemaFilterKey );
+            String[] schemasFilterArray = schemasFilterCommaList.split( "," );
+            for ( int i = 0; i < schemasFilterArray.length; i++ ) {
+              schemaList.add( new Schema( schemasFilterArray[i].trim() ) );
+            }
+          }
+          if ( schemaList.size() == 0 ) {
+            // Support schemas for MS SQL server due to PDI-1531
+            //
+            String sql = databaseMeta.getSQLListOfSchemas();
+            if ( !Utils.isEmpty( sql ) ) {
+              Statement schemaStatement = db.getConnection().createStatement();
+              ResultSet schemaResultSet = schemaStatement.executeQuery( sql );
+              while ( schemaResultSet != null && schemaResultSet.next() ) {
+                String schemaName = schemaResultSet.getString( "name" );
+                schemaList.add( new Schema( schemaName ) );
+              }
+              schemaResultSet.close();
+              schemaStatement.close();
+            } else {
+              ResultSet schemaResultSet = databaseMeta.getDatabaseInterface().getSchemas( dbmd, databaseMeta );
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
Analysis of this package determined that this line of code contains a resource that might not have closed properly. A resource leak can slow down or crash your system.

**Fix**
Consider closing the following resource in a try-finally block: *schemaResultSet*. The resource is referenced at lines 315, 316. The resource is closed at line 321. There are other execution paths that do not contain closure statements, for example, when ResultSet.next throws an exception.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).